test/igetest.c

```diff
@@ -145,129 +151,113 @@ static struct bi_ige_test const bi_ige_test_vectors[] = {
+        testresult = 0;
     }
 
-    return errs;
+    return testresult;
 }
 
-int main(int argc, char **argv)
+static int test_bi_ige_vectors(int n)
 {
-    unsigned char rkey[16];
-    unsigned char rkey2[16];
-    AES_KEY key;
+    const struct bi_ige_test *const v = &bi_ige_test_vectors[n];
+    AES_KEY key1;
     AES_KEY key2;
-    unsigned char plaintext[BIG_TEST_SIZE];
-    unsigned char ciphertext[BIG_TEST_SIZE];
-    unsigned char checktext[BIG_TEST_SIZE];
-    unsigned char iv[AES_BLOCK_SIZE * 4];
-    unsigned char saved_iv[AES_BLOCK_SIZE * 4];
-    int err = 0;
-    unsigned int n;
-    unsigned matches;
+    unsigned char buf[MAX_VECTOR_SIZE];
+
+        if (!TEST_int_ge(BIG_TEST_SIZE, TEST_SIZE)
+            || !TEST_int_le(v->length, MAX_VECTOR_SIZE))
+        return 0;
+
+    if (v->encrypt == AES_ENCRYPT) {
+        AES_set_encrypt_key(v->key1, 8 * v->keysize, &key1);
+        AES_set_encrypt_key(v->key2, 8 * v->keysize, &key2);
+    } else {
+        AES_set_decrypt_key(v->key1, 8 * v->keysize, &key1);
+        AES_set_decrypt_key(v->key2, 8 * v->keysize, &key2);
+    }
 
-    assert(BIG_TEST_SIZE >= TEST_SIZE);
+    AES_bi_ige_encrypt(v->in, buf, v->length, &key1, &key2, v->iv,
+                       v->encrypt);
 
-    RAND_bytes(rkey, sizeof rkey);
-    RAND_bytes(plaintext, sizeof plaintext);
-    RAND_bytes(iv, sizeof iv);
-    memcpy(saved_iv, iv, sizeof saved_iv);
+    if (!TEST_mem_eq(v->out, v->length, buf, v->length)) {
+        hexdump(stderr, "key 1", v->key1, sizeof v->key1);
+        hexdump(stderr, "key 2", v->key2, sizeof v->key2);
+        hexdump(stderr, "iv", v->iv, sizeof v->iv);
+        hexdump(stderr, "in", v->in, v->length);
+        return 0;
+    }
 
-    /* Forward IGE only... */
+    return 1;
+}
+
+static int test_ige_enc_dec(void)
+{
+    AES_KEY key;
+    unsigned char iv[AES_BLOCK_SIZE * 4];
+    unsigned char ciphertext[BIG_TEST_SIZE];
+    unsigned char checktext[BIG_TEST_SIZE];
+
+    if (!TEST_int_ge(BIG_TEST_SIZE, TEST_SIZE))
+        return 0;
 
-    /* Straight encrypt/decrypt */
+    memcpy(iv, saved_iv, sizeof iv);
     AES_set_encrypt_key(rkey, 8 * sizeof rkey, &key);
     AES_ige_encrypt(plaintext, ciphertext, TEST_SIZE, &key, iv, AES_ENCRYPT);
 
     AES_set_decrypt_key(rkey, 8 * sizeof rkey, &key);
     memcpy(iv, saved_iv, sizeof iv);
     AES_ige_encrypt(ciphertext, checktext, TEST_SIZE, &key, iv, AES_DECRYPT);
 
-    if (memcmp(checktext, plaintext, TEST_SIZE)) {
-        printf("Encrypt+decrypt doesn't match\n");
-        hexdump(stdout, "Plaintext", plaintext, TEST_SIZE);
-        hexdump(stdout, "Checktext", checktext, TEST_SIZE);
-        ++err;
-    }
+    return TEST_mem_eq(checktext, TEST_SIZE, plaintext, TEST_SIZE);
+}
+
+static int test_ige_enc_chaining(void)
+{
+    AES_KEY key;
+    unsigned char iv[AES_BLOCK_SIZE * 4];
+    unsigned char ciphertext[BIG_TEST_SIZE];
+    unsigned char checktext[BIG_TEST_SIZE];
+
+    if (!TEST_int_ge(BIG_TEST_SIZE, TEST_SIZE))
```

Comment:
It is a bit.  We've no way of failing a test in the `register_tests` functions.  This should also be added.  Once that is done, the bounds check can be done once before any tests are registered.